### PR TITLE
Sketch workbook implementation of the space emissions tool

### PR DESCRIPTION
**Pull request by kevin-hausmann**
Implementing the ideas from the wiki at https://github.com/UBA-DE-Emissionsituation/space-emissions/wiki and from issue #2.
Please feel free to comment and exchange thoughts/ideas.

How to test this:
1. Make sure to use a machine that has Jupyter Notebooks installed (or see hint below)
2. Get a local copy of this repo: `git clone https://github.com/UBA-DE-Emissionsituation/space-emissions.git`
3. Switch to this branch: `git checkout origin/sketch`
4. Open the notebook in the root folder and run all cells

A good online environment to test this actually comes with code-de.org, simply log into `https://jupyter.code-de.org/user/<user name>/lab` - should work great (use full email address as user name)!

### CODE REVIEW

**Comment by kevin-hausmann** [comment]
Hey @Christian-Mielke , @enricodammers and @jptokaya - I expanded this prototype quite a bit over the weekend, please review! I put some hints above on how to test this. Please comment if this all makes sense to you.
Also, please let me know if you have any preference/experience regarding the use of spatial libs in python. Should we use [GeoPandas](https://geopandas.org/) or [geojson](https://pypi.org/project/geojson)? Are .geojson files the way to go for us to represent the area the calculation should operate on? What object type should we give to the `EOEmissionCalculator.run()` method as an argument?

**Comment by jptokaya** [comment]
> Hey @Christian-Mielke , @enricodammers and @jptokaya - I expanded this prototype quite a bit over the weekend, please review! I put some hints above on how to test this. Please comment if this all makes sense to you.
> Also, please let me know if you have any preference/experience regarding the use of spatial libs in python. Should we use [GeoPandas](https://geopandas.org/) or [geojson](https://pypi.org/project/geojson)? Are .geojson files the way to go for us to represent the area the calculation should operate on? What object type should we give to the `EOEmissionCalculator.run()` method as an argument?

**Comment by kevin-hausmann** [comment]
Hi @jptokaya - why did you close this? Please feel free to test and comment, but this pull request is not yet merged (so it should not be closed). Reopening.

**Comment by jptokaya** [comment]
Sorry about that, Kevin. I closed it unintentionally. I'm still getting (re-)acquainted with working with git. I wanted to comment that I prefer GeoPandas, but must have hit "Close with comment"...

**Comment by kevin-hausmann** [comment]
Yes, I agree. I started looking into GeoPandas and I like what I see. What do you think would be the right object type to present to the `run()` method? GeoPandas seems to use `shapely.geometry.MultiPolygon` ...?

**Comment by jptokaya** [comment]
> Yes, I agree. I started looking into GeoPandas and I like what I see. What do you think would be the right object type to present to the `run()` method? GeoPandas seems to use `shapely.geometry.MultiPolygon` ...?

If my understanding is correct we only use numpy and regualar pandas for processing and some inhouse module called grid. For visualization purposes some people use GeoPandas. I mainly used basemap from matplotlib for displaying data. We can figure out along the way what best suits our purpose, or not? 

Btw we made a figure (as part of wp3) of the workflow as we think it can work: 
Processing stepsv2
Would it be handy to include this or an adaptation thereof in our workbook?

**Comment by kevin-hausmann** [comment]
> > Yes, I agree. I started looking into GeoPandas and I like what I see. What do you think would be the right object type to present to the `run()` method? GeoPandas seems to use `shapely.geometry.MultiPolygon` ...?
> 
> If my understanding is correct we only use numpy and regualar pandas for processing and some inhouse module called grid. For visualization purposes some people use GeoPandas. I mainly used basemap from matplotlib for displaying data. We can figure out along the way what best suits our purpose, or not?

Of course, you can use all the libs you want internally to your module. But, there has to be an interface between the overall tool and the module (one of potentially many) you are providing. I propose this interface here in the wiki at https://github.com/UBA-DE-Emissionsituation/space-emissions/wiki. Your module will have to implement the `run()` method as described there. One of the arguments for that method is the `area`, which has to be given as some data type. We could pass around paths to .geojson files, but I do not think that is the way to go. Therefore, if you agree, I would use `shapely.geometry.MultiPolygon` instead. What do you think?

Please note that the interface also strictly defines the return results for the `run()` method! Please have a look and let me know if that will work for you and if I am missing something!

> Btw we made a figure (as part of wp3) of the workflow as we think it can work:
> [snip]
> Would it be handy to include this or an adaptation thereof in our workbook?

Sure, that would be great. Please note that you can provide your own workbook for demonstration, but you do not have to do that, please focus on the implementation of the method itself.

**Comment by Christian-Mielke** [comment]
> > Yes, I agree. I started looking into GeoPandas and I like what I see. What do you think would be the right object type to present to the `run()` method? GeoPandas seems to use `shapely.geometry.MultiPolygon` ...?
> 
> If my understanding is correct we only use numpy and regualar pandas for processing and some inhouse module called grid. For visualization purposes some people use GeoPandas. I mainly used basemap from matplotlib for displaying data. We can figure out along the way what best suits our purpose, or not?
> 
> Btw we made a figure (as part of wp3) of the workflow as we think it can work:
> Processing stepsv2
> Would it be handy to include this or an adaptation thereof in our workbook?

Very nice and concise figure, I like this quite a lot.

File: eocalc/context.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    L_OtherWasteDisp = auto()
+    M_WasteWater = auto()
+    N_WasteIncin = auto()
+    O_AgriLivestock = auto()
+    P_AgriOther = auto()
+    Q_AgriWastes = auto()
+    R_Other = auto()
+    S_Natural = auto()
+    T_IntAviCruise = auto()
+    z_Memo = auto()
```
Comment (Christian-Mielke) [inline eocalc/context.py:45]: The pollutant classes are fine, but I think attributing to GNFR sectors would take things too far. Unless we would link/intersect/join our lps/prtr locations and values with this project here as well.

Comment (kevin-hausmann) [reply eocalc/context.py:45]: Hmm, I think this is actually quite handy to have. Note that there is no relation to the LPS project. I use this enum here:

```python
data = DataFrame(index=GNFR, columns=[f"{pollutant.name} [kt]", "Umin [%]", "Umax [%]"])
for sector in GNFR:
    data.loc[sector] = [random.random()*100, random.random()*18, random.random()*22]
# Add totals row at the bottom
data.loc["Totals"] = data.sum(axis=0)
```

Might come in handy for other methods as well...

**Comment by kevin-hausmann** [comment]
Addresses #2

**Pull request merged**